vowpalwabbit/cb_sample.cc

```diff
@@ -0,0 +1,62 @@
+#include "reductions.h"
+#include "cb_sample.h"
+#include "explore.h"
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+struct cb_sample_data
+{
+  uint64_t random_seed;
+  uint64_t random_seed_counter;
+};
+
+template <bool is_learn>
+void learn_or_predict(cb_sample_data& data, multi_learner& base, multi_ex& examples)
+{
+  multiline_learn_or_predict<is_learn>(base, examples, examples[0]->ft_offset);
+
+  auto action_scores = examples[0]->pred.a_s;
+
+  uint64_t seed = data.random_seed;
+  if (examples[0]->tag.size() > 0)
+  {
+    if (strncmp(examples[0]->tag.begin(), "seed=", 5) == 0)
+    {
+      substring tag_seed{examples[0]->tag.begin() + 5, examples[0]->tag.begin() + examples[0]->tag.size()};
+      seed = static_cast<uint64_t>(int_of_substring(tag_seed));
+    }
+  }
+
+  // Sampling is done after the base learner has generated a pdf.
+  uint32_t chosen_action;
+  if (exploration::sample_after_normalizing(seed + data.random_seed_counter++,
+          ACTION_SCORE::begin_scores(action_scores), ACTION_SCORE::end_scores(action_scores), chosen_action))
+    THROW("Failed to sample from pdf");
+
+  if (exploration::swap_chosen(action_scores.begin(), action_scores.end(), chosen_action))
+    THROW("Failed to swap top action_score chosen during sampling");
```

Comment:
Failure to swap seems like a silly reason to throw an exception.  Can we avoid this?

Reply:
I agree it doesn't make sense. But we need to explicitly handle all error that may come out of this function.

Reply:
Since this is on the example path, we shouldn't use an exception. 

Alternatives: 
1) Prove that the function can't err.
2) Define a default semantics when it errs.

Reply:
While I agree we need to be careful on the example path and this exception shouldn't be here. I think there are always going to be functions that can fail, and I don't think there is always a way to prevent that. 

I this case I believe a failure indicates it is totally broken, should we simply panic?

Reply:
Proving that we can't fail and then ignoring the possibility of failure seems more clean/efficient?

Reply:
Proving we can't fail seems a bit tricky, but the only sensible course of action here based on our policy.

Reply:
It still feels like a bug waiting to happen if we ignore error codes.

Reply:
We do other sanity checks in this file and throw if the invariants are not met, therefore I think leaving these here would be reasonable. They are a similar invariant check, if they are violated we cannot proceed. Thoughts?

Reply:
Maybe just use assert?  That way it has no runtime impact.

The general policy is no exceptions on the example handling path, which this certainly is.

Reply:
Assert is reasonable, I like that